src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,146 @@
+            tail = head;
+
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
+
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> nodeToInsert = new Node<>(null, value, null);
+
+        if (index == 0) {
+            nodeToInsert.next = head;
+            head.prev = nodeToInsert;
+            head = nodeToInsert;
+            size++;
+            return;
+        }
+
+        Node<T> current = getNodeByIndex(index);
+
+        nodeToInsert.prev = current.prev;
+        nodeToInsert.next = current;
+        current.prev.next = nodeToInsert;
+        current.prev = nodeToInsert;
+        size++;
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            return;
+        }
+        for (T element : list) {
+            add(element);
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeByIndex = getNodeByIndex(index);
+        return nodeByIndex.element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> nodeByIndex = getNodeByIndex(index);
+        T previousValue = nodeByIndex.element;
+        nodeByIndex.element = value;
+        return previousValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+
+        Node<T> current;
+        T element;
+
+        if (index == 0) {
+            element = head.element;
+            head = head.next;
+        } else {
+            current = getNodeByIndex(index);
+            element = current.element;
+            current.prev.next = current.next;
+
+            if (current.next != null) {
+                current.next.prev = current.prev;
+            }
+        }
+        size--;
+        return element;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (object == current.element || object != null && object.equals(current.element)) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; ++i) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid: ");
```

Comment:
maybe it will be better to add in the end value of the size or just correct the sentence without ":" in the end)